Zend/zend_alloc.c

```diff
@@ -660,28 +697,47 @@ static zend_always_inline int zend_mm_bitset_is_free_range(zend_mm_bitset *bitse
 
 static void *zend_mm_chunk_alloc_int(size_t size, size_t alignment)
 {
-	void *ptr = zend_mm_mmap(size);
-
+#ifdef _WIN32
+	void *ptr = NULL;
+	size_t offset;
+#ifndef _WIN64
+	/* 32-bit build has limited memory address space and partical deallocation is not supported,
+	 * try to reserve exact aligned size first, GH-9650 */
+	ptr = zend_mm_mmap_win_reserve(size + alignment);
 	if (ptr == NULL) {
 		return NULL;
-	} else if (ZEND_MM_ALIGNED_OFFSET(ptr, alignment) != 0) {
-		size_t offset;
+	}
 
-		/* chunk has to be aligned */
-		zend_mm_munmap(ptr, size);
-		ptr = zend_mm_mmap(size + alignment - REAL_PAGE_SIZE);
-#ifdef _WIN32
-		offset = ZEND_MM_ALIGNED_OFFSET(ptr, alignment);
-		zend_mm_munmap(ptr, size + alignment - REAL_PAGE_SIZE);
-		ptr = zend_mm_mmap_fixed((void*)((char*)ptr + (alignment - offset)), size);
-		offset = ZEND_MM_ALIGNED_OFFSET(ptr, alignment);
-		if (offset != 0) {
-			zend_mm_munmap(ptr, size);
+	offset = ZEND_MM_ALIGNED_OFFSET(ptr, alignment);
+	if (offset != 0) {
+		offset = alignment - offset;
+	}
+	zend_mm_munmap(ptr, size + alignment);
+	ptr = zend_mm_mmap_win_reserve_fixed((void*)((char*)ptr + offset), size);
+#endif
+	if (ptr == NULL) {
+		ptr = zend_mm_mmap_win_reserve(size + alignment);
```

Comment:
On Win64 you reserve "size+alignment" then commit "size". Is this intended?
I suppose this way, you reserve two times more memory when it's necessary.
See for example how jemalloc does the alignment https://github.com/jemalloc/jemalloc/blob/dev/src/pages.c#L203

Reply:
the lib/link you provided has the same issue, memory allocation can fail (after unmap and then map with fixed addr) and the app has to deal with it

with 32-build I can reproduce this very consistently in about 1 in 10k tests

Reply:
I hardly believe this is the right approach and jemalloc and others are wrong.

@cmb69 @weltling can you find some Microsoft expert, that can confirm that the following code is the right solution to allocate aligned memory blocks. (e.g. 2MB aligned block of size 2MB).
```c
void *ptr = VirtualAlloc(NULL, size + alignment, MEM_RESERVE, PAGE_READWRITE);
size_t offset = ((size_t)ptr) & (alignment - 1);
ptr = VirtualAlloc((void*)((char*)ptr + (alignment - offset)), size, MEM_COMMIT, PAGE_READWRITE);
```

Reply:
I don't think that I can find any expert, but from what I can tell, the proper solution to allocate an aligned memory block would be using [VirtualAlloc2](https://learn.microsoft.com/en-us/windows/win32/api/memoryapi/nf-memoryapi-virtualalloc2) (where the desired alignment can be specified). However, that API is only available as of Windows Server 2016/Windows 10, so can't be used by us (or at least we need a fallback).

So, in lack of that API, the code above seems to be the next best solution. It is actually similar to what we are doing now, although we currently don't split the reservation and the commit. Note that the `VirtualAlloc(NULL, …)` call is unlikely to give a 2MiB aligned memory address; the only guarantee is that it is aligned on the page size (which is 64KiB for me). So we do the re-allocation most of the time anyway to force the 2MiB alignment. There is likely a small performance gain if we don't commit the `NULL` allocation, because that zeroes the memory, while only reserving it won't touch the memory contents. The obvious drawback of splitting reservation/commit would be that we need to track the reserved address (or fetch it via `VirtualQuery()`).

I wonder, though, why we even try to enforce any alignment larger than the page size. From what I can tell, jemalloc would also do that, but it seems that they use a smaller page size than 2MiB (unless explicitly configured).

All in all, I see potential for improvement regarding the allocation, but it would make sense to actually do some performance measurements, but these experiments might better target the "master" branch only.

Reply:
I did more research and official Microsoft mimalloc is using exactly the same approach.

aligned alloc reserve: https://github.com/microsoft/mimalloc/blob/v2.0.6/src/os.c#L766 and commit: https://github.com/microsoft/mimalloc/blob/v2.0.6/src/os.c#L776

free https://github.com/microsoft/mimalloc/blob/v2.0.6/src/os.c#L366

Reply:
OK. Now I see that you are not alone who use this trick. (I'm not sure if it good or not. jemalloc is more known/used library and it uses simple loop in case of failure. Personally, I would prefer jemalloc version).

BTW: You do the trick only for Win32 (not Win64), but call VirtulQuery for both Win32 and Win64. Right? It's useless on Win64. Right?

Reply:
reserve more + commit only aligned size is used on both x86/x64

on x86, the original approach (subject to race condition) is tried once first to try to reserve the address space more densely, on x64 not needed and will imply only more syscalls

please note `#ifdef _WIN32` matches x86/x64 and `#ifdef _WIN32` x64 only

Reply:
> please note #ifdef _WIN32 matches x86/x64 and #ifdef _WIN64 x64 only

Right, but also for ARM.

Reply:
> I wonder, though, why we even try to enforce any alignment larger than the page size. From what I can tell, jemalloc would also do that, but it seems that they use a smaller page size than 2MiB (unless explicitly configured).

I was looking into the allocator last year or so - it's because of https://github.com/php/php-src/blob/98bdb7f99b60d4976797e9764c1c265413ab7915/Zend/zend_alloc.c#L40-L45

Given a pointer returned by emalloc, you can figure out which 2MiB chunk (or the address of the chunk header) as part of by a single bitwise operation 

https://github.com/php/php-src/blob/98bdb7f99b60d4976797e9764c1c265413ab7915/Zend/zend_alloc.c#L203-L225

Reply:
Right. Sorry, I was confused by ``#ifndef _WIN64`` block in zend_mm_chunk_alloc_int().
Why it's necessary only for 32-bit Windows? (actually, it implements the jemalloc approach)

Reply:
mvorisek mentioned why he'd went with this here https://github.com/php/php-src/pull/9721#issuecomment-1277890477

In practice, it sounds like 64-bit Windows doesn't have this issue, and the VirtualFree+VirtualAlloc within the freed region works consistently on 64-bit windows but not 32-bit windows

(I don't have windows installed to test)

> > The approach of a finite number of retries I'd mentioned earlier seems like the best option.
>
> several tries of "alloc + dealloc + smaller aligned alloc" fails once the 1st failed, that is almost expected, we will have to store the allocated parts first... and still only a probabilistic "almost never to fail" solution

Reply:
I don't understand. 
- Win32 first tries jemalloc approcah resevrce(sizealigned)+free+aligned_reserve(size)+commit,
then fallback to reserve(size+aligned)+aligned_commit(size),
- Win64 only the second fallback reserve(size+aligned)+aligned_commit(size)
Why Win64 don't use the jemalloc approch (it should work). Or why Win32 use it, if it doesn't work?

Reply:
I think this ``#ifndef _WIN64`` should be removed. Only #ifndef itself (not the code below).

Reply:
in 32-bit Windows the virtual memory space is limited to 2GB, so we try the racy allocation first which might fail and fallback to the fixed/x64 allocation, but if it succeed, only the exact memory range is reserved - in 64-bit, we do not need it, we can simply reserve more (and still commit only the aligned exact size)

this PR is not subject to any race condition (for both x86/x64), fixes all the memory random failures and I did not notice any performance nor other problems

also, similar solution is used by very reputable mimalloc memory allocator which is developer directly by Microsoft and has very good performance characteristics

Reply:
> I think this `#ifndef _WIN64` should be removed. Only #ifndef itself (not the code below).

`#ifndef _WIN64` means not 64-bit Windows, ie. 32-bit Windows only (if wrapped in `#ifdef _WIN32`), I double checked the code, the `#ifndef _WIN64` seems correct

Reply:
yes, Why this code should work only for 32-bit Windows? This means we will waste address space on Win64 only. 

Do you have race condition problems on ZTS build only?

Reply:
Look, you have a problem on 32-bit Windows, and you fix this adding fallback code for 32-bit Windows, but you also completely change the behavior for 64-bit Windows to use new fallback instead of old approach.

Reply:
we will waste a few atoms in universe (64-bit virtual address space) :D - is that a problem? bigger reservation should be fine, we only reserve larger range, but the actual commited physical memory size is always the same/exact

as long as there is no measureable problem (and I do not know any), I prefer to not even try the racy approach first for x64, as it needs more syscalls + error handling using global `GetLastError()`

ZTS - no, I can reproduce the problem consistently also with NTS

Reply:
I can't understand this. This "racy approach" worked for ages, and you tell it fails with 0.1% probability on Win32 only (I've never seen this kind of failures). Why do you change Win64 related behavior, and do it in a way that is really hard to catch. This patch is not just a fix!

Can you rewrite the fix that only adds the new fallback (I don't care for win32 only or for both). And do it in a compact readable way?

Reply:
problem is present on x86 and x64, on x64 it is much more rare as the addr collisions are much more rare

the original "racy" approach is not needed at all, it is there only to try to reserve exact memory size first to not reduce limited 2GB address space

so simply the original "racy" approach is problematic more or less on all Windows

> can you rewrite the fix that only adds the new fallback (I don't care for win32 only or for both). And do it in a compact readable way?

the approach in this PR uses minimal number of syscalls, it cannot be simplified

I would say this path is a fix, as the original approach is wrong/unfixable, no public API is changed and the allocator still uses the same `VirtualAlloc` function.

`#ifndef _WIN64` can be removed in theory, but I prefer to not do so, as it would only increase the number of syscalls, reserving "size+aligned" in 64-bit address space should not be worse than "size" only - is this the problem you have with this PR? If yes, I propose to remove the `#ifndef _WIN64` and add them for `master`.

Reply:
I gave up. I can't waste more time. I don't approve this, but fell free to commit if others approve. I won't object. (May be I'm a stubborn donkey).

Reply:
>I can't understand this. This "racy approach" worked for ages, and you tell it fails with 0.1% probability on Win32 only (I've never seen this kind of failures).

It's an issue I've been seeing in a small fraction of ci builds, when starting around a thousand php processes total per 32-bit execution (in a plugin starting a lot of php processes for `--syntax-check`, where each build included 2 jobs using 32-bit php for the test suites of the linked project)

https://ci.appveyor.com/project/TysonAndre/phan/builds/44983646/job/1f74k3nsurgxpdgb (more build failures in the job history, all jobs were running 32 bit on windows because I hadn't gotten around to setting up 32 bits on Linux before that

```
Analyzing files...
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░
Can't initialize heap: [0x000001e7] Attempt to access invalid address
░░░░░  54 / 619 (  8%) 409MB
░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ 108 / 619 ( 17%) 426MB
```

After seeing the cause of the failure I migrated AppVeyor to use 64 bit php and added 32-bit linux tests to work around this https://github.com/phan/phan/commit/35f0e26e49ced83dc1fd8e9fbf771f60dd927e3a

Reply:
> Can you rewrite the fix that only adds the new fallback (I don't care for win32 only or for both). And do it in a compact readable way?

I agree with that review comment.

I would have gone with it if Dmitry Stogov could tell if it was correct and agreed with the approach, but I'm not confident in this (e.g. CI tests are different from a long running web server).

- My earlier suggestion of keeping the method order unchanged and adding C function declarations would make this smaller and easier to review

It seems safer and easier to maintain to change only on affected platforms  to avoid unexpectedly changing behavior of this (e.g. to guarantee it only affects Windows), especially in 8.1, which already has stable releases



The null checks seem beneficial to ensure the error codes are from the first syscall that failed, but smaller PRs would be easier to review or bisect if anything did go wrong

Reply:
@TysonAndre thank you for rephrasing what I need to do ❤️

I will update this PR to be a pure fallback with minimal LOC changes needed.

For master, I will then submit another PR with the current impl. which is not a "fallback" (for non-x86) but is optimized to require less syscalls.

Reply:
done